modules/infra/datapath/loop_input.c

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include <gr.h>
+#include <gr_control_input.h>
+#include <gr_graph.h>
+#include <gr_ip4_datapath.h>
+#include <gr_ip6_datapath.h>
+#include <gr_loopback.h>
+#include <gr_trace.h>
+
+#include <rte_graph_worker.h>
+
+static control_input_t control_to_loopback_input;
+
+control_input_t loopback_get_control_id(void) {
+	return control_to_loopback_input;
+}
+
+enum {
+	IP_LOCAL,
+	IP_OUTPUT,
+	IP6_LOCAL,
+	IP6_OUTPUT,
+	IP_NO_ROUTE,
+	IP6_NO_ROUTE,
+	EDGE_COUNT,
+};
+
+static uint16_t loopback_input_process(
+	struct rte_graph *graph,
+	struct rte_node *node,
+	void **objs,
+	uint16_t nb_objs
+) {
+	struct rte_mbuf *mbuf;
+	rte_edge_t edge;
+
+	for (unsigned i = 0; i < nb_objs; i++) {
+		mbuf = control_input_mbuf_data(objs[i])->data;
+
+		if (gr_mbuf_is_traced(objs[i])
+		    || mbuf_data(mbuf)->iface->flags & GR_IFACE_F_PACKET_TRACE) {
+			// control input may already have added a trace.
+			// Copy it and clear it so we do not display it twice.
+			struct __mbuf_data *priv_old = rte_mbuf_to_priv(objs[i]);
+			struct __mbuf_data *priv = rte_mbuf_to_priv(mbuf);
+			priv->traces = priv_old->traces;
+			STAILQ_INIT(&priv_old->traces);
+			gr_mbuf_trace_add(mbuf, node, 0);
+		}
+		rte_pktmbuf_free(objs[i]);
```

Comment:
Why do you need to free the packet before sending it?

Reply:
fixed with the updated control to stack mechanism.